### PR TITLE
LB-232: Show bar graph for most listened to artists for users

### PR DESCRIPTION
**Pull request by paramsingh**
This is a bit of a work in progress because I'm not really good at d3 yet.

TODO:

- [x] Modularize the graph creation code 
- [x] Fail more graciously if no stats are preset
- [ ] Link MBIDs to MusicBrainz
- [ ] Link the artist page from somewhere (currently it isn't linked anywhere)
- [x] Add tests

Screen:

screenshot_2017-10-22_23-14-53

### CODE REVIEW

**Comment by mayhem** [comment]
Do you want me to look at this now? Do you need particular feedback?

**Comment by paramsingh** [comment]
K, this should be open for review now, I've created a few tickets as to future work that needs to be done here: [LB-254](https://tickets.metabrainz.org/browse/LB-254) and [LB-255](https://tickets.metabrainz.org/browse/LB-255), but those seem like bigger tasks to me and this has been open for too long. I'll do them after this gets reviewed.

File: listenbrainz/db/stats.py [hunk 1]
```diff
@@ -45,29 +65,69 @@ def insert_user_stats(user_id, artists, recordings, releases, artist_count):
...
+    """Get top artists for user with given ID.
+
+        Args:
+            user_id (int): the row ID of the user in the DB
+
+        Returns:
+            A dict of the following format
+            {
+                'user_id' (int): the row ID of the user in the DB,
+                'artist'  (dict): artist stats for the user
```
Comment (mayhem) [inline listenbrainz/db/stats.py]: We should be more specific what the stats are that are returned. What do they look like?

File: listenbrainz/db/stats.py [hunk 2]
```diff
@@ -45,29 +65,69 @@ def insert_user_stats(user_id, artists, recordings, releases, artist_count):
...
+    """ Get ALL user stats for user with given ID.
+
+        Args:
+            user_id (int): the row ID of the user in the DB
+
+        Returns:
+            A dict of the following format
+            {
+                'user_id' (int): the row ID of the user in the DB
+                'artist' (dict): artist stats for the user
```
Comment (mayhem) [inline listenbrainz/db/stats.py]: Same here -- if they all fit a common pattern, maybe describe this pattern somewhere?

File: listenbrainz/webserver/views/user.py [hunk 3]
```diff
@@ -139,6 +136,28 @@ def profile(user_name):
...
+    """ Show the top artists for the user. These users must have been already
+        calculated using Google BigQuery. If the stats are not present, we
+        redirect to the user page with a message.
+    """
+    user = _get_user(user_name)
+    data = db_stats.get_user_artists(user.id)
+
+    # if no data, flash a message and return to profile page
+    if data is None:
+        flash.error('No data calculated for user {user_name} yet.'.format(user_name=user_name))
```
Comment (mayhem) [inline listenbrainz/webserver/views/user.py]: I really think we need to improve this error message. If someone sees this, can they expect to wait a few seconds for the data to appear? If not, can they do anything to get the data to appear?

File: listenbrainz/webserver/views/user.py [hunk 4]
```diff
@@ -139,6 +136,28 @@ def profile(user_name):
...
 
 
+@user_bp.route("/<user_name>/artists")
+def artists(user_name):
+    """ Show the top artists for the user. These users must have been already
+        calculated using Google BigQuery. If the stats are not present, we
+        redirect to the user page with a message.
+    """
+    user = _get_user(user_name)
+    data = db_stats.get_user_artists(user.id)
```
Comment (mayhem) [inline listenbrainz/webserver/views/user.py]: And you're missing error handling. Again. I think we need to add a new tick box to the PR submission form:

[ ] Yes, I think I added enough error checking.

:)

Comment (paramsingh) [reply listenbrainz/webserver/views/user.py]: Sorry! Will keep in mind going forward. :(

Comment (eshansingh) [reply listenbrainz/webserver/views/user.py]: @mayhem @paramsingh Hey, I actually thought this was a pretty good idea, so I made a small PR for it. See #311.

Comment (mayhem) [reply listenbrainz/webserver/views/user.py]: LOL, funny. :)

**Comment by paramsingh** [comment]
I came across some discrepancies in how stats were being
stored in each of the three columns of the stat.user table,
leading to some misleading usage in the tests as well.
I've fixed that in this commit, now each stat has the
same format, i.e. a dict of the following form
```
{
    stat1: value1,
    stat2: value2,
    ...
}
```
See commit https://github.com/metabrainz/listenbrainz-server/pull/279/commits/94e6d76f240f6fb530e0829d8439607d928e842e

**Pull request merged**